server/service/src/invoice/inbound_shipment/add_from_master_list.rs

```diff
@@ -0,0 +1,307 @@
+use crate::invoice::common::{
+    AddToShipmentFromMasterListError as ServiceError,
+    AddToShipmentFromMasterListInput as ServiceInput,
+};
+use crate::invoice::{check_invoice_exists, common::check_master_list_for_name};
+use crate::service_provider::ServiceContext;
+use crate::sync_processor::invoice::common::get_lines_for_invoice;
+use repository::EqualFilter;
+use repository::{
+    InvoiceLine, InvoiceLineFilter, InvoiceLineRepository, InvoiceLineRow,
+    InvoiceLineRowRepository, InvoiceRow, InvoiceRowStatus, InvoiceRowType, MasterListLineFilter,
+    MasterListLineRepository, RepositoryError, StorageConnection,
+};
+
+use super::generate_unallocated_invoice_lines;
+
+pub fn add_from_master_list(
+    ctx: &ServiceContext,
+    store_id: &str,
+    input: ServiceInput,
+) -> Result<Vec<InvoiceLine>, ServiceError> {
+    let invoice_lines = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let invoice_row = validate(connection, store_id, &input)?;
+            let new_invoice_line_rows = generate(ctx, invoice_row, &input)?;
+
+            let invoice_line_row_repository = InvoiceLineRowRepository::new(&connection);
+
+            for invoice_line_row in new_invoice_line_rows {
+                invoice_line_row_repository.upsert_one(&invoice_line_row)?;
+            }
+
+            match InvoiceLineRepository::new(connection).query_by_filter(
+                InvoiceLineFilter::new().invoice_id(EqualFilter::equal_to(&input.shipment_id)),
+            ) {
+                Ok(lines) => Ok(lines),
+                Err(error) => Err(ServiceError::DatabaseError(error)),
+            }
+        })
+        .map_err(|error| error.to_inner_error())?;
+    Ok(invoice_lines)
+}
+
+fn validate(
+    connection: &StorageConnection,
+    store_id: &str,
+    input: &ServiceInput,
+) -> Result<InvoiceRow, ServiceError> {
+    let invoice_row = match check_invoice_exists(&input.shipment_id, connection) {
+        Ok(row) => row,
+        Err(_error) => return Err(ServiceError::ShipmentDoesNotExist),
+    };
+
+    if invoice_row.store_id != store_id {
+        return Err(ServiceError::NotThisStoreShipment);
+    }
+    if invoice_row.status == InvoiceRowStatus::Shipped
```

Comment:
check_invoice_is_eidtable would be good to use here (similar in outbound shipment). Btw we've been avoiding the use of `from` in the error mapping (after we've done this invoices though, they have the `from` mapping). So the syntax would be something like this:
```
check_invoice_is_editable(&invoice_row).map_err(|| ServiceError::CannotEditShipment)
```

technically speaking check_invoice_is_editable should return a boolean, but would require too much refactor

Reply:
Actually, ignore above please, for:
## Inbound Shipment

looks like we trim empty lines when status goes from `new` to any other status, which would mean we only allow to add from master list when status is `new`, otherwise trimming won't happen if we added from master list in other then new status), which means your logic here is correct (except it could just be invoice_row.status != InvoiceRowStatus::New). Are we allowed to add from master list for transfers ? This would mean status == Shipped

## Outbound Shipment

We should also check similar, it looks like we trim after 'new' status but adding from master list is allowed for 'Allocated' (at which point we would have unallocated lines in allocated status, and lines would not be trimmed)